### ISSUE TITLE
DateSet type and primitive operations

### ISSUE THREAD

**Issue by amosr** [ambiata/icicle#12]
We want an efficient, specialised type that corresponds to the MutableDateSet in Scala.
We can only create these things for constant, known sizes so we can have the type parameterised by an int, the size.
I think in Core we can get away with requiring the size to be specified upfront, even though it should be inferred from the window.

Note that the evaluation and whatnot don't have to be efficient, but the point is that specialising code generation for this case will be easy.

tldr
new DateSet type (DateSetT Int)
primitives for DateSets in Core.Prim

**Comment by amosr** [comment]
Also worth noting that this is only really necessary for the quantiles - mean days, max days and so on currently use the Scala MutableDateSet, but I believe we should be able to get away with a single fold for those cases.

**Comment by amosr** [comment]
I am convinced this is no longer necessary: most cases where the MutableDateSet is used in the Ivory implementation can actually be implemented as a single fold without a set (see #49), and for the few other cases, I think we are better off focusing on making a better general Map.
